src/coins.cpp

```diff
@@ -49,10 +49,8 @@ CCoinsMap::iterator CCoinsViewCache::FetchCoin(const COutPoint &outpoint) const
             cacheCoins.erase(ret);
             return cacheCoins.end();
         }
-        if (ret->second.coin.IsSpent()) {
-            // The parent only has an empty entry for this outpoint; we can consider our version as fresh.
-            ret->second.AddFlags(CCoinsCacheEntry::FRESH, *ret, m_sentinel);
-        }
+        // GetCoin will never return true if the coin is spent
+        Assume(!ret->second.coin.IsSpent());
```

Comment:
It would be a fatal error when running in DEBUG mode, but is not compiled in for release.

> when do we throw and when do we Assume here?

I only added throws to `BatchWrite` because there are invalid states that can happen in our fuzz test harness which we want to handle properly. A follow up to remove setting the flags directly can help make that more clear and we wouldn't need to have the throws there anymore.

Reply:
Split out to https://github.com/bitcoin/bitcoin/pull/30906